### PR TITLE
[VL] Add sort merge join metrics

### PR DESCRIPTION
**Pull request by ulysses-you**
## What changes were proposed in this pull request?

This pr adds metrics for sort merge join. Since join always has post project, unfiy the post project metrics for hash join and sort merge join.

## How was this patch tested?

Add test and test manually

image

### CODE REVIEW

File: backends-velox/src/main/scala/io/glutenproject/backendsapi/velox/MetricsApiImpl.scala [hunk 1]
```diff
@@ -476,20 +496,13 @@ class MetricsApiImpl extends MetricsApi with Logging {
       "postProjectionWallNanos" -> SQLMetrics.createNanoTimingMetric(
         sparkContext,
         "totaltime of postProjection"),
-      "postProjectionOutputRows" -> SQLMetrics.createMetric(
-        sparkContext,
-        "number of postProjection output rows"),
-      "postProjectionOutputVectors" -> SQLMetrics.createMetric(
-        sparkContext,
-        "number of postProjection output vectors"),
-      "finalOutputRows" -> SQLMetrics.createMetric(sparkContext, "number of final output rows"),
```
Comment (ulysses-you) [inline backends-velox/src/main/scala/io/glutenproject/backendsapi/velox/MetricsApiImpl.scala:485]: it's nerver used before, and replaced to numOutputRows in this pr

**Comment by ulysses-you** [comment]
cc @PHILO-HE @JkSelf thank you

File: backends-velox/src/test/scala/io/glutenproject/execution/VeloxMetricsSuite.scala [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+  }
+
+  override protected def afterAll(): Unit = {
+    spark.sql("drop table metrics_t1")
+    spark.sql("drop table metrics_t2")
+
+    super.afterAll()
+  }
+
+  test("test sort merge join metrics") {
```
Comment (JkSelf) [inline backends-velox/src/test/scala/io/glutenproject/execution/VeloxMetricsSuite.scala:54]: @ulysses-you Do we need to add the test with post project? And also add the hash join test for both with pre project and post project?

**Review by JkSelf — COMMENTED** [review]
@ulysses-you Thanks for your great work. Just two small questions.

File: backends-velox/src/main/scala/io/glutenproject/backendsapi/velox/MetricsApiImpl.scala [hunk 3]
```diff
@@ -352,12 +352,32 @@ class MetricsApiImpl extends MetricsApi with Logging {
...
+      "streamPreProjectionCpuCount" -> SQLMetrics.createMetric(
+        sparkContext,
+        "stream preProject cpu wall time count"),
+      "streamPreProjectionWallNanos" -> SQLMetrics.createNanoTimingMetric(
+        sparkContext,
+        "totaltime of stream preProjection"),
+      "bufferPreProjectionCpuCount" -> SQLMetrics.createMetric(
+        sparkContext,
+        "buffer preProject cpu wall time count"),
+      "bufferPreProjectionWallNanos" -> SQLMetrics.createNanoTimingMetric(
```
[the remark this answers is not in the document]
Comment (ulysses-you) [reply backends-velox/src/main/scala/io/glutenproject/backendsapi/velox/MetricsApiImpl.scala:372]: Merge join only supports inner and left outer, so the steam side is always left child and the buffer side is always right child. It's used to show the metrics pre-project for stream/buffer side.

File: backends-velox/src/test/scala/io/glutenproject/execution/VeloxMetricsSuite.scala (hunk 2 above)
Comment (ulysses-you) [reply backends-velox/src/test/scala/io/glutenproject/execution/VeloxMetricsSuite.scala:54]: Join always has the post project. The numOutputRows/numOutputVectors metrics are from post project.

**Review by JkSelf — APPROVED** [review]
LGTM. Thanks.

**Pull request merged**